File: 백준/Silver/1018. 체스판 다시 칠하기/체스판 다시 칠하기.java
import java.io.*;
import java.util.*;

public class Main {
    public static boolean[][] array;
    public static int min = 64;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken()), M = Integer.parseInt(st.nextToken());

        array = new boolean[N][M];

        for(int i = 0; i < N; i++){
            String str = br.readLine();

            for(int j = 0; j < M; j++){
                if(str.charAt(j) == 'W'){ // 흰색이면 true, 검은색이면 false로 array에 삽입
                    array[i][j] = true;
                } else {
                    array[i][j] = false;
                }
            }

        }

        br.close();

        // 경우의수는 N-7, M-7 (예: 8*8은 1*1로 한가지이다.)
        int cntN = N - 7;
        int cntM = M - 7;

        for(int i = 0; i < cntN; i++){
            for(int j = 0; j < cntM; j++){
                check(i, j);
            }
        }
        System.out.println(min);
    }

    public static void check(int x, int y){
        int rowLimit = x + 8;
        int colLimit = y + 8;
        int cnt = 0;

        boolean checker = array[x][y]; // 시작하는 타일의 색깔 ('W' = true, 'B' = false)

        for(int i = x; i < rowLimit; i++){
            for(int j = y; j < colLimit; j++){

                if(array[i][j] != checker){ // 타일 색깔이 맞지 않으면 교체카운트++
                    cnt++;
                }

                checker = (!checker); // 다음 타일은 색깔이 바뀌므로 변경해주기

            }

            checker = (!checker); // 다음 타일은 색깔이 바뀌므로 변경해주기
        }

        cnt = Math.min(cnt, 64-cnt); // 시작색깔 기준 카운트와 아닌색깔 기준 카운트 중 더 작은 값을 cnt로 초기화
        min = Math.min(cnt, min);
    }
}

File: 백준/Silver/1018. 체스판 다시 칠하기/체스판 다시 칠하기.kt
import java.io.*
import java.util.*
import kotlin.math.min

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val st = StringTokenizer(readLine())
    val N = st.nextToken().toInt()
    val M = st.nextToken().toInt()

    // B, W 여부 담기 위한 boolean 배열
    val arr = Array(N) {BooleanArray(M)}

    for(i in 0 until N) {
        val input = readLine()

        for(j in 0 until M) {
            if(input[j] == 'W') { // W면 true
                arr[i][j] = true
            } else {
                arr[i][j] = false // B면 false
            }
        }
    }

    close()
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    // 8 X 8 배열이기 때문에 값 64로 초기화
    var minVal = 64

    for(i in 0 until N-7) {
        for(j in 0 until M-7) {
            minVal = min(minVal, check(arr, i, j))
        }
    }

    bw.write(minVal.toString())
    bw.flush()
    bw.close()
}

fun check(arr : Array<BooleanArray>, x : Int, y : Int) : Int {
    var isWhite = arr[x][y]
    var cnt = 0 // 색깔 다시 칠할 정사각형 갯수

    for(i in x until x + 8) {
        for(j in y until y + 8) {

            // 현재 위치 값이 이전과 같으면 색깔을 다시 칠해야 함
            if(isWhite != arr[i][j]) {
                cnt++
            }

            // column 위치가 변경되면 색깔이 바뀜
            isWhite = !isWhite
        }

        // row가 바뀔 때마다 또 한번 색깔이 바뀜
        isWhite = !isWhite
    }

    return min(cnt, 64 - cnt)
}